File: typescript/src/features/claude_code/slash_commands.ts
// Re-implementation of Claude Code's Slash Command Router

export type CommandHandler = (args: string[]) => Promise<string>;

export class SlashCommandRouter {
  private commands: Map<string, CommandHandler> = new Map();

  public registerCommand(command: string, handler: CommandHandler): void {
    if (!command.startsWith('/')) {
      throw new Error("Slash commands must start with '/'");
    }
    this.commands.set(command, handler);
  }

  public isSlashCommand(input: string): boolean {
    return input.trim().startsWith('/');
  }

  public async executeCommand(input: string): Promise<string> {
    const parts = input.trim().split(/\s+/);
    const cmd = parts[0];
    const args = parts.slice(1);

    const handler = this.commands.get(cmd);
    if (!handler) {
      return `Unknown command: ${cmd}. Type /help for available commands.`;
    }

    return await handler(args);
  }
}

File: rust/src/claude_code/slash_commands.rs
// Re-implementation of Claude Code's Slash Command Router

use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;

pub type CommandHandler = Box<dyn Fn(Vec<String>) -> Pin<Box<dyn Future<Output = String> + Send>> + Send + Sync>;

pub struct SlashCommandRouter {
    commands: HashMap<String, CommandHandler>,
}

impl SlashCommandRouter {
    pub fn new() -> Self {
        Self {
            commands: HashMap::new(),
        }
    }

    pub fn register_command(&mut self, command: &str, handler: CommandHandler) -> Result<(), String> {
        if !command.starts_with('/') {
            return Err("Slash commands must start with '/'".to_string());
        }
        self.commands.insert(command.to_string(), handler);
        Ok(())
    }

    pub fn is_slash_command(&self, input: &str) -> bool {
        input.trim().starts_with('/')
    }

    pub async fn execute_command(&self, input: &str) -> String {
        let mut parts = input.trim().split_whitespace();
        if let Some(cmd) = parts.next() {
            let args: Vec<String> = parts.map(String::from).collect();
            if let Some(handler) = self.commands.get(cmd) {
                handler(args).await
            } else {
                format!("Unknown command: {}. Type /help for available commands.", cmd)
            }
        } else {
            "Empty command.".to_string()
        }
    }
}
